Autodetect path for assemble and run scripts

anmolbabu opened the pull request:
Autodetect path for assemble and run scripts

Odo hardcodes the S2I assemble script path in
`https://github.com/redhat-developer/odo-supervisord-image/blob/master/assemble-and-restart`
But the S2I docs suggest that the S2I scripts need to looked up in the
following order:
1. A script found at the --scripts-url URL
2. A script found in the application source .s2i/bin directory
3. A script found at the default image URL (io.openshift.s2i.scripts-url label)
and ultimately giveup if not found in any of the above paths.

While, 1 is not relevant to us(builds triggered internally by odo), 2 is already
done by odo, 3 is enhanced by this PR instead of using the currently(without this PR)
hard-coded path

This PR fetches the S2I scripts paths from builder image and extracts the protocol and path
and sets them into environment variables in component container for use of it in running
assemble script.

Note: The S2I scripts suggest that the builder image might point to any of the following
paths:

a. image://path_to_scripts_dir - absolute path inside the image
b. file://path_to_scripts_dir - relative or absolute path on the host machine
c. http(s)://path_to_scripts_dir - URL to a directory

Of these, the assemble-and-restart script currently due to:
`https://github.com/redhat-developer/odo-supervisord-image/pull/3` is capable of handling
`image://path_to_scripts_dir` but not `file://` or `http(s)://` and they are expected to be
added over time. At the moment for any of the `file://` or `http(s)://`, the component
create and/or update is expected to fail as it is expected to be even with latest master.

To test:
Using minishift:
1. Start minishift `minishift start`
2. Build the docker image for `odo-supervisord-image` with PR:
        `https://github.com/redhat-developer/odo-supervisord-image/pull/3`.
   To do this follow the below steps:
   i. To pull the PR:
      * Clone `https://github.com/redhat-developer/odo-supervisord-image`
      * Add my git remote: `git remote add anmolbabu git@github.com:anmolbabu/odo-supervisord-image.git`
      * Pull source from the remote `git pull anmolbabu`
      * Checkout my branch `git checkout remotes/anmolbabu/fixes-odo#869`
      * Create new branch for the new source `git checkout -b anmolbabu-fixes-odo#869`
  ii. Get access to minishift docker environment `eval $(minishift docker-env)`
 iii. Build the `odo-supervisord-image` from the above branch may be with a new version(that's not in quay)
      to be sure we are using the image with the required PRs:
      `docker build -t quay.io/openshiftdo/supervisord:0.3.0 .`
  iv. Now using the odo binary for the PR `https://github.com/redhat-developer/odo/pull/924`(using either the
      one generated on the PR by travis or built locally using steps similar to 1 above), ensure `odo create`
      followed by `odo push` and `odo update` works...

fixes: #869
Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

anmolbabu commented:
ToDo(@anmolbabu ): 
1. Fix UTs
2. Verify PR

anmolbabu commented on pkg/occlient/occlient_test.go:
```diff
@@ -3651,6 +3643,167 @@ func TestUpdateDCToGit(t *testing.T) {
 	}
 }
 
+func TestuniqueAppendOrOverwriteEnvVars(t *testing.T) {
```
But this is not required to be exported to outside just to maintain the Camel case conventions...

cdrage commented on pkg/occlient/occlient.go:
```diff
@@ -949,11 +1063,30 @@ func (c *Client) UpdateDCToSupervisor(commonObjectMeta metav1.ObjectMeta, compon
 
 	// Add the appropriate bootstrap volumes for SupervisorD
 	addBootstrapVolumeCopyInitContainer(&dc, commonObjectMeta.Name)
+	//addBootstrapSupervisordInitContainer(&dc, commonObjectMeta.Name)
```
Don't we need this? Why are we removing the bootstrap supervisord image?

anmolbabu replied:
Well I am not very sure why I added this line and commented it out because it’s already there in the line below and is exactly the same...😀

anmolbabu commented:
@cdrage

> You removed adding the bootstrap image, but for some reason the tests didn't fail.. 
Anyways! Here's my review :+1: 

No not really because there’s the same code statement again in the line below... So, I think I had just added that line and commented it out 😀

anmolbabu commented on pkg/occlient/occlient.go:
```diff
@@ -546,6 +547,21 @@ func (c *Client) GetExposedPorts(imageStream *imagev1.ImageStream, imageTag stri
+func (c *Client) GetExposedPorts(imageStreamImage *imagev1.ImageStreamImage) ([]corev1.ContainerPort, error) {
+	var containerPorts []corev1.ContainerPort
+
+	glog.V(4).Infof("Checking for exact match of builderImage with ImageStream")
```
The comment can actually be removed safely now as its no more relevant there :)

anmolbabu commented on pkg/occlient/occlient.go:
```diff
@@ -635,6 +656,71 @@ func (c *Client) NewAppS2I(commonObjectMeta metav1.ObjectMeta, builderImage stri
+		mapExistingEnvs[envVar.Name] = envVar
+	}
+
+	// For each new envVar to be appended, Add(if envVar with same name doesn't already exist)/overwrite(if envVar with same name already exists) the map
```
Ack

anmolbabu commented on pkg/occlient/occlient.go:
```diff
@@ -635,6 +656,71 @@ func (c *Client) NewAppS2I(commonObjectMeta metav1.ObjectMeta, builderImage stri
+		return "", "", fmt.Errorf("Unknown scripts url %s", s2iScriptsURL)
+	}
+	return s2iScriptsProtocol, s2iScriptsPath, nil
+}
```
Ack

anmolbabu commented on pkg/occlient/occlient.go:
```diff
@@ -635,6 +656,71 @@ func (c *Client) NewAppS2I(commonObjectMeta metav1.ObjectMeta, builderImage stri
+	}
+
+	var dimdr DockerImageMetaDataRaw
+	dimdrByteArr := (*builderImage).Image.DockerImageMetadata.Raw
```
Ack

anmolbabu commented:
@cdrage Incorporated your comments Please revisit it

cdrage commented:
FYI: The `sample.war` file I am using is from our tests: https://gist.github.com/mik-dass/f95bd818ddba508ff76a386f8d984909/raw/e5bc575ac8b14ba2b23d66b5cb4873657e1a1489/sample.war

cdrage commented:
Full logs:

```sh
github.com/redhat-developer/odo  pr_924 ✔                                                                                                                                                                                                                                                                                                                           3h40m  ⍉
▶ ./odo create wildfly --binary ~/sample.war && ./odo push -v 4        
Component 'sample-wildfly-rflg' was created and port 8080/TCP was opened
To push source code to the component run 'odo push'

Component 'sample-wildfly-rflg' is now set as active component.
I1106 15:38:42.499530   32457 occlient.go:288] Trying to connect to server 192.168.42.19:8443
I1106 15:38:42.499815   32457 occlient.go:294] Server https://192.168.42.19:8443 is up
I1106 15:38:42.508522   32457 occlient.go:260] isLoggedIn err:  <nil>
 output: "developer"
Pushing changes to component: sample-wildfly-rflg
I1106 15:38:42.511341   32457 occlient.go:2124] Getting DeploymentConfig: sample-wildfly-rflg-odo-mnlo
I1106 15:38:42.528754   32457 component.go:479] Source for component sample-wildfly-rflg is file:///home/wikus/sample.war (binary)
I1106 15:38:42.528775   32457 push.go:83] Copying file /home/wikus/sample.war to pod
I1106 15:38:42.534510   32457 occlient.go:1268] Waiting for deploymentconfig=sample-wildfly-rflg-odo-mnlo pod
I1106 15:38:45.097974   32457 occlient.go:1290] Status of sample-wildfly-rflg-odo-mnlo-1-8b2lf pod is Pending
I1106 15:38:45.102215   32457 occlient.go:1290] Status of sample-wildfly-rflg-odo-mnlo-1-8b2lf pod is Pending
I1106 15:38:45.188740   32457 occlient.go:1290] Status of sample-wildfly-rflg-odo-mnlo-1-8b2lf pod is Pending
I1106 15:38:46.920106   32457 occlient.go:1290] Status of sample-wildfly-rflg-odo-mnlo-1-8b2lf pod is Pending
I1106 15:38:47.959647   32457 occlient.go:1290] Status of sample-wildfly-rflg-odo-mnlo-1-8b2lf pod is Pending
I1106 15:38:50.123262   32457 occlient.go:1290] Status of sample-wildfly-rflg-odo-mnlo-1-8b2lf pod is Running
I1106 15:38:50.123278   32457 occlient.go:1293] Pod sample-wildfly-rflg-odo-mnlo-1-8b2lf is running.
I1106 15:38:50.123307   32457 component.go:346] Copying to pod sample-wildfly-rflg-odo-mnlo-1-8b2lf
E1106 15:41:37.582539   32457 v2.go:105] write tcp 192.168.42.1:57360->192.168.42.19:8443: use of closed network connection
I1106 15:41:37.582609   32457 client.go:38] Error:
unable push files to pod: error while streaming command: Internal error occurred: error executing command in container: read unix @->/var/run/docker.sock: read: connection reset by peer
failed to push component: sample-wildfly-rflg
```

geoand commented:
@anmolbabu @cdrage @kadel I'm looking into the Wildfly issue. We might need to utilize another one of the labels (I think there is a label that indicates where binaries are stored) in order to get it to work

anmolbabu commented:
ToDo: https://github.com/redhat-developer/odo/blob/master/pkg/component/component.go#L326

geoand commented:
@anmolbabu See my various comments in odo-supervisord-image/pull/3

I proposed a change [there](https://github.com/redhat-developer/odo-supervisord-image/pull/3#discussion_r231792343) that suggest to odo source code as well :)

cdrage commented:
Are you able to reproduce this error @anmolbabu ?

It seems that I'm unable to deploy the wildfly binary:
```sh
github.com/redhat-developer/odo  pr_924 ✔                                                                                                                                    0h40m  ⍉
▶ ./odo create wildfly wildfly --binary ~/sample.war
Component 'wildfly' was created and port 8080/TCP was opened
To push source code to the component run 'odo push'

Component 'wildfly' is now set as active component.

github.com/redhat-developer/odo  pr_924 ✔                                                                                                                                     0h40m
▶ ./odo push -v 4
I1108 12:03:27.398039   12483 occlient.go:308] Trying to connect to server 192.168.42.155:8443
I1108 12:03:27.398392   12483 occlient.go:314] Server https://192.168.42.155:8443 is up
I1108 12:03:27.410737   12483 occlient.go:280] isLoggedIn err:  <nil>
 output: "developer"
Pushing changes to component: wildfly
I1108 12:03:27.412723   12483 occlient.go:2306] Getting DeploymentConfig: wildfly-odo-pftp
I1108 12:03:27.424018   12483 component.go:478] Source for component wildfly is file:///home/wikus/sample.war (binary)
I1108 12:03:27.424079   12483 push.go:83] Copying file /home/wikus/sample.war to pod
I1108 12:03:27.432061   12483 occlient.go:1436] Waiting for deploymentconfig=wildfly-odo-pftp pod
I1108 12:03:27.439437   12483 occlient.go:1458] Status of wildfly-odo-pftp-1-kgb88 pod is Pending
I1108 12:05:21.519600   12483 occlient.go:1458] Status of wildfly-odo-pftp-1-kgb88 pod is Pending
I1108 12:05:22.544398   12483 occlient.go:1458] Status of wildfly-odo-pftp-1-kgb88 pod is Pending
I1108 12:05:23.584613   12483 occlient.go:1458] Status of wildfly-odo-pftp-1-kgb88 pod is Running
I1108 12:05:23.584627   12483 occlient.go:1461] Pod wildfly-odo-pftp-1-kgb88 is running.
I1108 12:05:23.584650   12483 component.go:345] Copying to pod wildfly-odo-pftp-1-kgb88
E1108 12:07:38.426971   12483 occlient.go:2395] Error while creating tar: &os.PathError{Op:"open", Path:"/home/wikus/.minikube/machines/minikube/boot2docker.iso", Err:0xd}
```

cdrage commented:
Otherwise, this works correctly on NodeJS / Ruby / Python!

geoand commented:
> Are you able to reproduce this error @anmolbabu ?
> 
> It seems that I'm unable to deploy the wildfly binary:
> 
> ```shell
> github.com/redhat-developer/odo  pr_924 ✔                                                                                                                                    0h40m  ⍉
> ▶ ./odo create wildfly wildfly --binary ~/sample.war
> Component 'wildfly' was created and port 8080/TCP was opened
> To push source code to the component run 'odo push'
> 
> Component 'wildfly' is now set as active component.
> 
> github.com/redhat-developer/odo  pr_924 ✔                                                                                                                                     0h40m
> ▶ ./odo push -v 4
> I1108 12:03:27.398039   12483 occlient.go:308] Trying to connect to server 192.168.42.155:8443
> I1108 12:03:27.398392   12483 occlient.go:314] Server https://192.168.42.155:8443 is up
> I1108 12:03:27.410737   12483 occlient.go:280] isLoggedIn err:  <nil>
>  output: "developer"
> Pushing changes to component: wildfly
> I1108 12:03:27.412723   12483 occlient.go:2306] Getting DeploymentConfig: wildfly-odo-pftp
> I1108 12:03:27.424018   12483 component.go:478] Source for component wildfly is file:///home/wikus/sample.war (binary)
> I1108 12:03:27.424079   12483 push.go:83] Copying file /home/wikus/sample.war to pod
> I1108 12:03:27.432061   12483 occlient.go:1436] Waiting for deploymentconfig=wildfly-odo-pftp pod
> I1108 12:03:27.439437   12483 occlient.go:1458] Status of wildfly-odo-pftp-1-kgb88 pod is Pending
> I1108 12:05:21.519600   12483 occlient.go:1458] Status of wildfly-odo-pftp-1-kgb88 pod is Pending
> I1108 12:05:22.544398   12483 occlient.go:1458] Status of wildfly-odo-pftp-1-kgb88 pod is Pending
> I1108 12:05:23.584613   12483 occlient.go:1458] Status of wildfly-odo-pftp-1-kgb88 pod is Running
> I1108 12:05:23.584627   12483 occlient.go:1461] Pod wildfly-odo-pftp-1-kgb88 is running.
> I1108 12:05:23.584650   12483 component.go:345] Copying to pod wildfly-odo-pftp-1-kgb88
> E1108 12:07:38.426971   12483 occlient.go:2395] Error while creating tar: &os.PathError{Op:"open", Path:"/home/wikus/.minikube/machines/minikube/boot2docker.iso", Err:0xd}
> ```

I never experienced this... I'll need to try again tomorrow

kadel commented:
> Are you able to reproduce this error @anmolbabu ?
> 
> It seems that I'm unable to deploy the wildfly binary:
> 
> ```shell
> github.com/redhat-developer/odo  pr_924 ✔                                                                                                                                    0h40m  ⍉
> ▶ ./odo create wildfly wildfly --binary ~/sample.war
> Component 'wildfly' was created and port 8080/TCP was opened
> To push source code to the component run 'odo push'
> 
> Component 'wildfly' is now set as active component.
> 
> github.com/redhat-developer/odo  pr_924 ✔                                                                                                                                     0h40m
> ▶ ./odo push -v 4
> I1108 12:03:27.398039   12483 occlient.go:308] Trying to connect to server 192.168.42.155:8443
> I1108 12:03:27.398392   12483 occlient.go:314] Server https://192.168.42.155:8443 is up
> I1108 12:03:27.410737   12483 occlient.go:280] isLoggedIn err:  <nil>
>  output: "developer"
> Pushing changes to component: wildfly
> I1108 12:03:27.412723   12483 occlient.go:2306] Getting DeploymentConfig: wildfly-odo-pftp
> I1108 12:03:27.424018   12483 component.go:478] Source for component wildfly is file:///home/wikus/sample.war (binary)
> I1108 12:03:27.424079   12483 push.go:83] Copying file /home/wikus/sample.war to pod
> I1108 12:03:27.432061   12483 occlient.go:1436] Waiting for deploymentconfig=wildfly-odo-pftp pod
> I1108 12:03:27.439437   12483 occlient.go:1458] Status of wildfly-odo-pftp-1-kgb88 pod is Pending
> I1108 12:05:21.519600   12483 occlient.go:1458] Status of wildfly-odo-pftp-1-kgb88 pod is Pending
> I1108 12:05:22.544398   12483 occlient.go:1458] Status of wildfly-odo-pftp-1-kgb88 pod is Pending
> I1108 12:05:23.584613   12483 occlient.go:1458] Status of wildfly-odo-pftp-1-kgb88 pod is Running
> I1108 12:05:23.584627   12483 occlient.go:1461] Pod wildfly-odo-pftp-1-kgb88 is running.
> I1108 12:05:23.584650   12483 component.go:345] Copying to pod wildfly-odo-pftp-1-kgb88
> E1108 12:07:38.426971   12483 occlient.go:2395] Error while creating tar: &os.PathError{Op:"open", Path:"/home/wikus/.minikube/machines/minikube/boot2docker.iso", Err:0xd}
> ```

Based on your log I think that this is caused by https://github.com/redhat-developer/odo/issues/950.
You are pushing war file that is in your HOME. Because of that bug, it tries to push your whole HOME :-) It probably fails because it is too big. Even from time, you can see that tar creation failed after 2 minutes :-D

kadel commented:
It looks like python is not working.
```
▶ odo push -v 4
I1108 20:44:41.392357   12823 occlient.go:308] Trying to connect to server 192.168.99.100:8443
I1108 20:44:41.392876   12823 occlient.go:314] Server https://192.168.99.100:8443 is up
I1108 20:44:41.412506   12823 occlient.go:280] isLoggedIn err:  <nil>
 output: "developer"
Pushing changes to component: mypython
I1108 20:44:41.412865   12823 occlient.go:2306] Getting DeploymentConfig: mypython-mysqlpython-dxna
I1108 20:44:41.423209   12823 component.go:478] Source for component mypython is file:///Users/tomas/tmp/odo-examples/mysqlpython (local)
I1108 20:44:41.423254   12823 push.go:79] Copying directory /Users/tomas/tmp/odo-examples/mysqlpython to pod
I1108 20:44:41.439762   12823 occlient.go:1436] Waiting for deploymentconfig=mypython-mysqlpython-dxna pod
I1108 20:44:41.444952   12823 occlient.go:1458] Status of mypython-mysqlpython-dxna-1-8q2x6 pod is Running
I1108 20:44:41.444970   12823 occlient.go:1461] Pod mypython-mysqlpython-dxna-1-8q2x6 is running.
I1108 20:44:41.445004   12823 component.go:345] Copying to pod mypython-mysqlpython-dxna-1-8q2x6
Please wait, building component....
+ set -eo pipefail
+ '[' -f /.s2i/bin/assemble ']'
+ '[' -n /usr/libexec/s2i ']'
+ rm -rf /opt/app-root/src/.git
+ '[' '' '!=' /opt/app-root/src ']'
+ cp -a /opt/app-root/src/
cp: missing destination file operand after ‘/opt/app-root/src/’
Try 'cp --help' for more information.
I1108 20:44:41.662701   12823 client.go:38] Error:
unable to execute assemble script: error while streaming command: command terminated with exit code 1
failed to push component: mypython
```

I'm going to investigate ...

geoand commented:
Can you share the project you are using to test?

kadel commented:
> Can you share the project you are using to test?

it is my test script that I used to test service and env injection, it it just two files -  https://gist.github.com/kadel/5c0e832fe06285565a88ac0f1c5c6162

kadel commented on pkg/occlient/occlient.go:
```diff
@@ -681,6 +718,102 @@ func secretKeyName(componentName, baseKeyName string) string {
+
+	if s2iSrcOrBinPath == "" {
+		// In cases like nodejs builder image, where there is no concept of binary and sources are directly run, use destination as source
+		s2iSrcOrBinPath = getS2ILabelValue(dimdr.ContainerConfig.Labels, S2IDeploymentsDir)
```
This is not right. If `io.openshift.s2i.destination` is not set, it should use `/tmp` by default. See https://github.com/openshift/source-to-image/blob/master/docs/builder_image.md#required-image-contents

anmolbabu replied:
Its seems not to be the case for nodejs however...
I'll confirm once my minishift is back to normal 
I see that my system /var/ has filled up as in :
```
[root@localhost redhat]# du -h -d 1 /var/
4.0K	/var/games
4.0K	/var/adm
4.0K	/var/local
516K	/var/tmp
4.0K	/var/crash
24K	/var/www
4.0K	/var/gopher
4.0K	/var/ftp
4.0K	/var/nis
4.0K	/var/account
21G	/var/lib
43G	/var/cache
12K	/var/kerberos
4.2G	/var/log
452K	/var/spool
8.0K	/var/empty
4.0K	/var/yp
4.0K	/var/preserve
12K	/var/db
4.0K	/var/opt
67G	/var/

```

geoand replied:
Perhaps you need to do `minishift delete` and create a new VM :).

NodeJS does indeed work

geoand replied:
I assume that `/tmp` should be the absolute last resort @kadel ?

anmolbabu replied:
Incorporated now

kadel commented on pkg/occlient/occlient.go:
```diff
@@ -681,6 +718,102 @@ func secretKeyName(componentName, baseKeyName string) string {
+}
+
+// GetS2IPathsFromBuilderImg returns script path protocol, S2I scripts path, S2I source or binary expected path, S2I deployment dir and errors(if any) from the passed builder image
+func GetS2IPathsFromBuilderImg(builderImage *imagev1.ImageStreamImage) (string, string, string, string, error) {
```
It would be better to remove struct than 5 independent strings. It is hard to know what value is what.

anmolbabu replied:
Ack
Done

kadel commented:
> I'm not sure that we handle `io.openshift.s2i.destination` correctly.
> 
> If I understand it correctly, `io.openshift.s2i.destination` is a directory where `assemble` source code.
> 
> That means that we should use that path everywhere where we have `/opt/app-root/`.
> 
> * in DeploymentConfig we should have volume for `io.openshift.s2i.destination` instead of `/opt/app-root`  (I think that we name that volume as `something-s2idata` or something like that)
> * `odo push` should also to `io.openshift.s2i.destination`
> * Am I forgetting something?

I'm ok with splitting this into two PRs, to avoid large changes. We came make changes to volumes and push in different PR that comes after this one. But the default `/tmp` needs to be fixed here, otherwise, it is broken for some builder images.

geoand commented:
I just want people to make sure that when they are testing this, they manually build the supervisord image according to @anmolbabu 's instructions:

```
Create new branch for the new source git checkout -b anmolbabu-fixes-odo#869
ii. Get access to minishift docker environment eval $(minishift docker-env)
iii. Build the odo-supervisord-image from the above branch may be with a new version(that's not in quay)
to be sure we are using the image with the required PRs:
docker build -t quay.io/openshiftdo/supervisord:0.3.0 .
```

but then they also need to change the supevisord image being used in odo:

`bootstrapperImage = "quay.io/openshiftdo/supervisord:0.2.0"`

Alternatively when building the docker image just build version `0.2.0`.

I am mentioning this because it's very easy to not perform one of these steps which results in testing the old version. I know it's happened to me :wink:

geoand commented:
> It looks like python is not working.
> 
> ```
> ▶ odo push -v 4
> I1108 20:44:41.392357   12823 occlient.go:308] Trying to connect to server 192.168.99.100:8443
> I1108 20:44:41.392876   12823 occlient.go:314] Server https://192.168.99.100:8443 is up
> I1108 20:44:41.412506   12823 occlient.go:280] isLoggedIn err:  <nil>
>  output: "developer"
> Pushing changes to component: mypython
> I1108 20:44:41.412865   12823 occlient.go:2306] Getting DeploymentConfig: mypython-mysqlpython-dxna
> I1108 20:44:41.423209   12823 component.go:478] Source for component mypython is file:///Users/tomas/tmp/odo-examples/mysqlpython (local)
> I1108 20:44:41.423254   12823 push.go:79] Copying directory /Users/tomas/tmp/odo-examples/mysqlpython to pod
> I1108 20:44:41.439762   12823 occlient.go:1436] Waiting for deploymentconfig=mypython-mysqlpython-dxna pod
> I1108 20:44:41.444952   12823 occlient.go:1458] Status of mypython-mysqlpython-dxna-1-8q2x6 pod is Running
> I1108 20:44:41.444970   12823 occlient.go:1461] Pod mypython-mysqlpython-dxna-1-8q2x6 is running.
> I1108 20:44:41.445004   12823 component.go:345] Copying to pod mypython-mysqlpython-dxna-1-8q2x6
> Please wait, building component....
> + set -eo pipefail
> + '[' -f /.s2i/bin/assemble ']'
> + '[' -n /usr/libexec/s2i ']'
> + rm -rf /opt/app-root/src/.git
> + '[' '' '!=' /opt/app-root/src ']'
> + cp -a /opt/app-root/src/
> cp: missing destination file operand after ‘/opt/app-root/src/’
> Try 'cp --help' for more information.
> I1108 20:44:41.662701   12823 client.go:38] Error:
> unable to execute assemble script: error while streaming command: command terminated with exit code 1
> failed to push component: mypython
> ```
> I'm going to investigate ...
> UPDATE1:
> default python builder image doesn't have `io.openshift.s2i.destination` label
> 
> UPDATE2:
> This PR is not using any default value for for `io.openshift.s2i.destination`, if not set it should use `/tmp`

I can also reproduce this and am seeing the exact behavior @kadel mentions in his updates.

kadel commented on pkg/occlient/occlient.go:
```diff
@@ -681,6 +721,115 @@ func secretKeyName(componentName, baseKeyName string) string {
+	return ""
+}
+
+type S2IPaths struct {
```
It is good practice to have all the definitions at the beginning of the file. It makes it easier to find them.

anmolbabu replied:
Ack
Done now

kadel commented on pkg/occlient/occlient.go:
```diff
@@ -74,8 +74,46 @@ const (
+	EnvS2IDeploymentDir = "ODO_S2I_DEPLOYMENT_DIR"
+
+	// DefaultS2ISrcOrBinPath is the default path where S2I expects source/binary artifacts in absence of $S2ISrcOrBinLabel in builder image
+	DefaultS2ISrcOrBinPath = "/tmp/src"
```
I believe that this should be set to `/tmp`. adding `src` should be handled on the cluster side.
Otherwise, we will have problems. As I understand it. s2i is actualy expecting source code to be in `io.openshift.s2i.destination`/src not directly in `io.openshift.s2i.destination`. That means that if image has `io.openshift.s2i.destination` set to `/my/dir` s2i expects source to be in  `/my/dir/src`.

geoand replied:
When we tried `/tmp` pushing source code for a python sample failed with the following logs:

```
Pushing changes to component: python-python-buxz
Please wait, building component....
+ set -eo pipefail
+ '[' -f /tmp/.s2i/bin/assemble ']'
+ '[' -n /usr/libexec/s2i ']'
+ rm -rf /opt/app-root/src/.git
+ '[' /tmp '!=' /opt/app-root/src ']'
+ cp -a /opt/app-root/src/ /tmp
+ /usr/libexec/s2i/assemble
---> Installing application source ...
mv: inter-device move failed: ‘/tmp/src/.pki’ to ‘./.pki’; unable to remove target: Directory not empty
```

anmolbabu replied:
Ack

cdrage commented:
Tested against Python, WildFly, NodeJS (with binary too) and Ruby and this implementation works for me :+1:

anmolbabu commented:
@kadel @geoand @metacosm @cdrage Please review

There have a few changes since last review and they have been verified for:
wildfly(binary+local), nodejs(local), openjdk(binary+local), python(local) 

Cc: @surajnarwade @mik-dass @syamgk

anmolbabu commented:
The test failure is bcoz there is no image available with https://github.com/redhat-developer/odo-supervisord-image/pull/3

kadel commented:
> The test failure is bcoz there is no image available with [redhat-developer/odo-supervisord-image#3](https://github.com/redhat-developer/odo-supervisord-image/pull/3)

when we merge the image, we will release it as  v0.3.0. It might be a good change image in this PR to this version. So we can verify that all tests are passing as soon as we release new image version

anmolbabu commented:
Yay :man_dancing: 3 acks on the pr
